src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -6271,229 +6271,205 @@ private TypeSyntax ParsePointerTypeMods(TypeSyntax type)
-                StatementSyntax result = ParseLocalDeclarationStatement();
-
-                // Case (1)
-                if (result == null)
-                {
-                    this.Reset(ref resetPointBeforeStatement);
-                    return null;
-                }
-
-                // Cases (2), (3) and (4):
-                if (!beginsWithAwait || !result.ContainsDiagnostics)
-                {
-                    return result;
-                }
+                return this.ParseExpressionStatement();
+            }
 
-                // The statement begins with "await" and could not be parsed as a legal declaration statement.
-                // We know from our precondition that it is not a legal "await X();" statement, though it is
-                // possible that it was only not legal because we were not in an async context.
+            if (isGlobalScriptLevel)
+            {
+                // if we're at the global script level, then we don't support local-decls or
+                // local-funcs. The caller instead will look for those and parse them as
+                // fields/methods in the global script scope.
+                return null;
+            }
 
-                Debug.Assert(!IsInAsync);
+            bool beginsWithAwait = this.CurrentToken.ContextualKind == SyntaxKind.AwaitKeyword;
+            var result = ParseLocalDeclarationStatement();
 
-                // Let's see if we're in case (5). Pretend that we're in an async method and retry.
+            // didn't get any sort of statement.  This was something else entirely
+            // (like just a `}`).  No need to retry anything here.  Just reset back
+            // to where we started from and bail entirely from parsing a statement.
+            if (result == null)
+            {
+                this.Reset(ref resetPointBeforeStatement);
+                return null;
+            }
 
+            if (result.ContainsDiagnostics &&
+                beginsWithAwait &&
+                !IsInAsync)
+            {
+                // Local decl had issues.  We were also starting with 'await' in a non-async
+                // context. Retry parsing this as if we were in an 'async' context as it's much
+                // more likely that this was a misplace await-expr' than a local decl.
+                //
+                // The user will still get a later binding error about an await-expr in a non-async
+                // context.
                 this.Reset(ref resetPointBeforeStatement);
+
                 IsInAsync = true;
-                result = ParseStatementNoDeclaration(allowAnyExpression: false);
+                result = ParseExpressionStatement();
                 IsInAsync = false;
-
-                return result;
-            }
-            finally
-            {
-                this.Release(ref resetPointBeforeStatement);
             }
+
+            // Didn't want to retry as an `await expr`.  Just return what we actually
```

Comment:
This comment seems off, we use this return statement whether or not we reparsed as an await expr.